# Hyper Parameter Tuning
- hyper parameter : 모델 설정과 관련해 직접 지정할 수 있는 매개변수
- model parameter : 회귀계수(가중치), 절편 등 모델의 학습 대상이 되는 변수

### GridSearchCV

In [2]:
from sklearn.datasets import load_iris
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

# 데이터 로드
iris_input, iris_target = load_iris(return_X_y=True)

# 모델 생성
knn = KNeighborsClassifier()

# 테스트할 파라미터 값
params = {
    'n_neighbors': range(1, 13, 2)
}

# 첫 번째 인자: 모델
# 두 번째 인자: 테스트 할 파라미터 (딕셔너리)
# scoring: 평가 지표 (accuracy, precision, recall, f1)
# cv: 반복 횟수
grid = GridSearchCV(knn, params, scoring='accuracy', cv=5)
grid.fit(iris_input, iris_target)

print("최적의 파라미터:", grid.best_params_)
print("최적의 모델 객체:", grid.best_estimator_)
print("최적화된 점수:", grid.best_score_)

최적의 파라미터: {'n_neighbors': 7}
최적의 모델 객체: KNeighborsClassifier(n_neighbors=7)
최적화된 점수: 0.9800000000000001


In [3]:
best_knn = grid.best_estimator_
best_knn.fit(iris_input, iris_target)
best_knn.score(iris_input, iris_target)

0.9733333333333334

### RandomSearchCV
- 하이퍼 파라미터의 값 목록이나 값의 범위를 제공하는데, 이 범위 중에 랜덤하게 값을 뽑아내 최적의 하이퍼 파라미터 조합을 찾는다.
    - 탐색범위가 넓을 때 짧은 시간 내에 좋은 결과를 얻을 수 있다.
    - 랜덤하게 값을 추출해 계산하므로, 전역 최적값을 놓칠 수 있다.

In [4]:
from sklearn.model_selection import RandomizedSearchCV

# 모델 생성
knn = KNeighborsClassifier()

# 테스트할 파라미터 생성
params = {
    'n_neighbors': range(1, 100, 2)
}

# n_iter: 탐색할 최적의 하이퍼 파라미터 조합 수 (기본값: 10)
#         값이 크면 시간이 오래 걸림 / 값이 작으면 좋은 조합을 찾을 가능성 저하 
rd_search = RandomizedSearchCV(knn, params, cv=5, n_iter=10, random_state=0)
rd_search.fit(iris_input, iris_target)

print("최적의 파라미터:", rd_search.best_params_)
print("최적의 모델 객체:", rd_search.best_estimator_)
print("최적화된 점수:", rd_search.best_score_)
rd_search.cv_results_

최적의 파라미터: {'n_neighbors': 5}
최적의 모델 객체: KNeighborsClassifier()
최적화된 점수: 0.9733333333333334


c:\Users\Playdata\anaconda3\envs\pystudy_env\Lib\site-packages\joblib\externals\loky\backend\context.py:136: UserWarning: Could not find the number of physical cores for the following reason:
[WinError 2] 지정된 파일을 찾을 수 없습니다
Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
  warnings.warn(
  File "c:\Users\Playdata\anaconda3\envs\pystudy_env\Lib\site-packages\joblib\externals\loky\backend\context.py", line 257, in _count_physical_cores
    cpu_info = subprocess.run(
               ^^^^^^^^^^^^^^^
  File "c:\Users\Playdata\anaconda3\envs\pystudy_env\Lib\subprocess.py", line 550, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\Playdata\anaconda3\envs\pystudy_env\Lib\subprocess.py", line 1028, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "c:\Users\Playdata\anaconda3\envs\pystudy_env\Lib\subproces

{'mean_fit_time': array([0.00090737, 0.00020437, 0.        , 0.00080085, 0.00068536,
        0.00039592, 0.00039973, 0.00020108, 0.00010824, 0.        ]),
 'std_fit_time': array([0.00018452, 0.00040874, 0.        , 0.00040047, 0.00041268,
        0.0004849 , 0.00048957, 0.00040216, 0.00021648, 0.        ]),
 'mean_score_time': array([0.00169344, 0.00179214, 0.0018084 , 0.01530046, 0.00120535,
        0.00176458, 0.00858521, 0.00829873, 0.00202346, 0.00176306]),
 'std_score_time': array([7.52403472e-04, 4.14124586e-04, 4.05897782e-04, 1.44086440e-02,
        4.05069408e-04, 3.54363689e-04, 1.05936218e-03, 1.46714670e-03,
        2.39821123e-05, 3.88762308e-04]),
 'param_n_neighbors': masked_array(data=[57, 23, 21, 83, 5, 55, 77, 63, 45, 9],
              mask=[False, False, False, False, False, False, False, False,
                    False, False],
        fill_value=999999),
 'params': [{'n_neighbors': 57},
  {'n_neighbors': 23},
  {'n_neighbors': 21},
  {'n_neighbors': 83},
  {'n_nei

---

### HyperOpt

**hyper.hp클래스**
<table border="1">
  <thead>
    <tr>
      <th>함수명</th>
      <th>설명</th>
      <th>사용 방법</th>
      <th>예시 코드</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>hp.uniform</td>
      <td>연속적인 실수 값 샘플링</td>
      <td>hp.uniform(label, low, high)</td>
      <td><code>hp.uniform('learning_rate', 0.01, 0.1)</code></td>
    </tr>
    <tr>
      <td>hp.quniform</td>
      <td>연속적이지만 일정 간격(q)을 갖는 값 샘플링</td>
      <td>hp.quniform(label, low, high, q)</td>
      <td><code>hp.quniform('num_layers', 1, 5, 1)</code></td>
    </tr>
    <tr>
      <td>hp.loguniform</td>
      <td>로그 스케일로 분포된 실수 값 샘플링</td>
      <td>hp.loguniform(label, low, high)</td>
      <td><code>hp.loguniform('reg_param', -3, 0)</code></td>
    </tr>
    <tr>
      <td>hp.randint</td>
      <td>정수 값 샘플링</td>
      <td>hp.randint(label, upper)</td>
      <td><code>hp.randint('num_trees', 1, 100)</code></td>
    </tr>
    <tr>
      <td>hp.choice</td>
      <td>주어진 리스트 중 임의의 값 샘플링</td>
      <td>hp.choice(label, options)</td>
      <td><code>hp.choice('optimizer', ['adam', 'sgd', 'rmsprop'])</code></td>
    </tr>
    <tr>
      <td>hp.normal</td>
      <td>정규분포에서 값 샘플링</td>
      <td>hp.normal(label, mean, std)</td>
      <td><code>hp.normal('dropout_rate', 0.3, 0.05)</code></td>
    </tr>
    <tr>
      <td>hp.lognormal</td>
      <td>로그 정규분포에서 값 샘플링</td>
      <td>hp.lognormal(label, mean, std)</td>
      <td><code>hp.lognormal('scale', 0, 1)</code></td>
    </tr>
  </tbody>
</table>

In [5]:
!pip install hyperopt

In [6]:
from hyperopt import hp

# 검색 공간
search_space = {
    'x': hp.quniform('x', -10, 10, 1),
    'y': hp.quniform('y', -15, 15, 1)
}

In [7]:
import hyperopt

# 목적 함수
def objective(search_space):
    x = search_space['x']
    y = search_space['y']

    return {
        'loss': x**2 + 20 * y,
        'status': hyperopt.STATUS_OK
    }

In [8]:
from hyperopt import fmin, tpe, Trials

# 탐색 과정을 저장하는 객체
trials = Trials()

# fmin() : 목적 함수의 최소값을 찾는 함수
best_val = fmin(
    fn=objective,           # 목적함수
    space=search_space,     # 검색공간
    algo=tpe.suggest,       # 베이지안 최적화 적용     
    max_evals=500,          # 반복 횟수
    trials=trials           # 탐색과정 저장
)

best_val

100%|██████████| 500/500 [00:05<00:00, 87.88trial/s, best loss: -300.0] 


{'x': np.float64(0.0), 'y': np.float64(-15.0)}

In [9]:
# 탐색과정 -> 목적함수 반환값 (loss와 실행 상태) 저장
trials.results

# 탐색과정 -> 하이퍼 파라미터값을 딕셔너리(리스트) 형태로 저장
trials.vals

{'x': [np.float64(3.0),
  np.float64(9.0),
  np.float64(2.0),
  np.float64(3.0),
  np.float64(6.0),
  np.float64(4.0),
  np.float64(4.0),
  np.float64(3.0),
  np.float64(4.0),
  np.float64(8.0),
  np.float64(6.0),
  np.float64(-6.0),
  np.float64(-9.0),
  np.float64(-8.0),
  np.float64(-8.0),
  np.float64(-3.0),
  np.float64(-3.0),
  np.float64(1.0),
  np.float64(-5.0),
  np.float64(3.0),
  np.float64(7.0),
  np.float64(-1.0),
  np.float64(-1.0),
  np.float64(-1.0),
  np.float64(-2.0),
  np.float64(-0.0),
  np.float64(-4.0),
  np.float64(0.0),
  np.float64(-6.0),
  np.float64(1.0),
  np.float64(-0.0),
  np.float64(-2.0),
  np.float64(1.0),
  np.float64(-5.0),
  np.float64(5.0),
  np.float64(10.0),
  np.float64(2.0),
  np.float64(2.0),
  np.float64(-1.0),
  np.float64(5.0),
  np.float64(-4.0),
  np.float64(-7.0),
  np.float64(-2.0),
  np.float64(9.0),
  np.float64(5.0),
  np.float64(4.0),
  np.float64(-10.0),
  np.float64(-4.0),
  np.float64(7.0),
  np.float64(-3.0),
  np.float64(-6.0),

- hyperopt를 활용한 XGBoost 하이퍼 파라미터 튜닝

In [10]:
from xgboost import XGBClassifier
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split, cross_val_score

data = load_breast_cancer()
X_train, X_test, y_train, y_test =\
    train_test_split(data.data, data.target, random_state=42)

# 1. 검색 공간
search_space = {
    'n_estimators': hp.quniform('n_estimators', 100, 500, 100),
    'max_depth': hp.quniform('max_depth', 3, 10, 1),
    'learning_rate': hp.uniform('learning_rate', 0.01, 0.2),
    'colsample_bytree': hp.uniform('colsample_bytree', 0.5, 1)
}

# 2. 목적 함수
def xgb_objective(ss):

    xgb_clf = XGBClassifier(
        n_estimators=int(ss['n_estimators']),
        max_depth=int(ss['max_depth']),
        learning_rate=ss['learning_rate'],
        colsample_bytree=ss['colsample_bytree']
    )
    mean_acc = cross_val_score(xgb_clf, X_train, y_train, scoring='accuracy', cv=3).mean()
    return {
        'loss': -1 * mean_acc,
        'status': hyperopt.STATUS_OK
    }


# 3. Trials() + fmin()
trials = Trials()
best = fmin(
    fn=xgb_objective,
    space=search_space,
    algo=tpe.suggest,
    max_evals=50,
    trials=trials
)
best

100%|██████████| 50/50 [00:20<00:00,  2.49trial/s, best loss: -0.971830985915493] 


{'colsample_bytree': np.float64(0.6931939616646488),
 'learning_rate': np.float64(0.1541783542739828),
 'max_depth': np.float64(8.0),
 'n_estimators': np.float64(400.0)}

---

### Optuna

<table border="1">
    <thead>
        <tr>
            <th>함수명</th>
            <th>설명</th>
            <th>사용 방법</th>
            <th>예시 코드</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>suggest_uniform</td>
            <td>연속적인 실수 값 샘플링</td>
            <td>trial.suggest_uniform(name, low, high)</td>
            <td><code>trial.suggest_uniform('learning_rate', 0.01, 0.1)</code></td>
        </tr>
        <tr>
            <td>suggest_discrete_uniform</td>
            <td>연속적이지만 일정 간격(step)을 갖는 값 샘플링</td>
            <td>trial.suggest_discrete_uniform(name, low, high, step)</td>
            <td><code>trial.suggest_discrete_uniform('num_layers', 1, 5, 1)</code></td>
        </tr>
        <tr>
            <td>suggest_loguniform</td>
            <td>로그 스케일로 분포된 실수 값 샘플링</td>
            <td>trial.suggest_loguniform(name, low, high)</td>
            <td><code>trial.suggest_loguniform('reg_param', 1e-3, 1)</code></td>
        </tr>
        <tr>
            <td>suggest_int</td>
            <td>정수 값 샘플링</td>
            <td>trial.suggest_int(name, low, high, step)</td>
            <td><code>trial.suggest_int('num_trees', 1, 100)</code></td>
        </tr>
        <tr>
            <td>suggest_categorical</td>
            <td>주어진 리스트 중 임의의 값 샘플링</td>
            <td>trial.suggest_categorical(name, choices)</td>
            <td><code>trial.suggest_categorical('optimizer', ['adam', 'sgd', 'rmsprop'])</code></td>
        </tr>
        <tr>
            <td>suggest_float</td>
            <td>연속적인 실수 값 샘플링 (<code>step</code> 사용 가능)</td>
            <td>trial.suggest_float(name, low, high, step=None, log=False)</td>
            <td><code>trial.suggest_float('alpha', 0.1, 1.0, step=0.1)</code></td>
        </tr>
    </tbody>
</table>

In [11]:
!pip install optuna

In [12]:
import optuna

# 목적 함수
def objective(trial):
    x = trial.suggest_uniform('x', -10, 10)
    y = trial.suggest_uniform('y', -15, 15)
    return (x -3) ** 2 + (y + 5) ** 2

# 스터디 생성
study = optuna.create_study(direction="minimize")

# 최적화 실행
study.optimize(objective, n_trials=500)
    
# 결과 확인
print(study.best_value)
print(study.best_params)

[I 2025-03-28 12:46:09,682] A new study created in memory with name: no-name-0d0efb53-fcf8-4456-b46e-bcbe7f465aac
C:\Users\Playdata\AppData\Local\Temp\ipykernel_1908\3508545321.py:5: FutureWarning: suggest_uniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float instead.
  x = trial.suggest_uniform('x', -10, 10)
C:\Users\Playdata\AppData\Local\Temp\ipykernel_1908\3508545321.py:6: FutureWarning: suggest_uniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float instead.
  y = trial.suggest_uniform('y', -15, 15)
[I 2025-03-28 12:46:09,684] Trial 0 finished with value: 7.935048465077573 and parameters: {'x': 5.7306924767922744, 'y': -4.308359152256283}. Best is trial 0 with value: 7.935048465077573.
[I 2025-03-28 12:46:09,685] Trial 1 finished with value: 452.27890850680484 and parameters: {'x': -7

0.0023991988060800135
{'x': 2.99979882435252, 'y': -4.951018796110764}


In [13]:
# !pip install plotly

In [14]:
import optuna.visualization as vis

# 하이퍼 파라미터 중요도 시각화
vis.plot_param_importances(study).show()

In [ ]:
# 최적화 히스토리 시각화
vis.plot_optimization_history(study).show()

- optuna를 활용한 XGBoost 하이퍼 파라미터 튜닝

In [17]:
# 1. 목적 함수
def xgb_optuna_objective(trial):
    params = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 500, 100),
        'max_depth': trial.suggest_int('max_depth', 3, 10),
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.2),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.5, 1.0)
    }
    xgb_clf = XGBClassifier(**params)
    return cross_val_score(xgb_clf, X_train, y_train, scoring='accuracy', cv=3).mean()


# 2. study 객체 -> 최적화
study = optuna.create_study(direction='maximize')
study.optimize(xgb_optuna_objective, n_trials=50)

# 3. 결과 출력
print(study.best_params)
print(study.best_value)

[I 2025-03-28 14:22:31,507] A new study created in memory with name: no-name-906917ed-2924-4ec2-8eee-ab7fae42112a
C:\Users\Playdata\AppData\Local\Temp\ipykernel_1908\1028312989.py:4: FutureWarning:

suggest_int() got {'step'} as positional arguments but they were expected to be given as keyword arguments.

[I 2025-03-28 14:22:32,003] Trial 0 finished with value: 0.9624413145539906 and parameters: {'n_estimators': 500, 'max_depth': 8, 'learning_rate': 0.021000054051863594, 'colsample_bytree': 0.7944320331334334}. Best is trial 0 with value: 0.9624413145539906.
C:\Users\Playdata\AppData\Local\Temp\ipykernel_1908\1028312989.py:4: FutureWarning:

suggest_int() got {'step'} as positional arguments but they were expected to be given as keyword arguments.

[I 2025-03-28 14:22:32,158] Trial 1 finished with value: 0.9624413145539906 and parameters: {'n_estimators': 200, 'max_depth': 8, 'learning_rate': 0.1341501845475387, 'colsample_bytree': 0.7354020834233754}. Best is trial 0 with value: 0.96

{'n_estimators': 500, 'max_depth': 9, 'learning_rate': 0.1659738245986902, 'colsample_bytree': 0.5046056593621138}
0.971830985915493


##### HyperOpt vs Optuna

- HyperOpt
    - 'colsample_bytree': np.float64(0.6931939616646488)
    - 'learning_rate': np.float64(0.1541783542739828)
    - 'max_depth': np.float64(8.0)
    - 'n_estimators': np.float64(400.0)

- Optuna
    - 'n_estimators': 500
    - 'max_depth': 9
    - 'learning_rate': 0.1659738245986902
    - 'colsample_bytree': 0.5046056593621138

In [18]:
from sklearn.metrics import accuracy_score

xgb_hpopt = XGBClassifier(
    n_estimators=400,
    max_depth=8,
    learning_rate=0.15,
    colsample_bytree=0.7
)

xgb_optuna = XGBClassifier(
    n_estimators=500,
    max_depth=9,
    learning_rate=0.17,
    colsample_bytree=0.5
)

xgb_hpopt.fit(X_train, y_train)
xgb_optuna.fit(X_train, y_train)

hpopt_pred = xgb_hpopt.predict(X_test)
optuna_pred = xgb_optuna.predict(X_test)

print(f"HyperOpt 최적 파라미터 적용: {accuracy_score(y_test, hpopt_pred)}")
print(f"Optuna 최적 파라미터 적용: {accuracy_score(y_test, optuna_pred)}")

HyperOpt 최적 파라미터 적용: 0.958041958041958
Optuna 최적 파라미터 적용: 0.958041958041958
